js/src/main/scala-2.10/scalajson.ast/JValue.scala

```diff
@@ -143,9 +142,85 @@ final class JNumber private[ast] (val value: String) extends JValue {
 
   def copy(value: String): JNumber =
     value match {
-      case jNumberRegex(_ *) => new JNumber(value)
+      case jNumberRegex(_ *) => new JNumber(value)(0)
       case _ => throw new NumberFormatException(value)
     }
+
+  def toInt: Option[Long] = {
+    if ((constructedFlag & NumberFlags.int) == NumberFlags.int)
+      Some(value.toInt)
+    else {
+      try {
+        val asInt = value.toInt
+        if (BigInt(value) == BigInt(asInt))
+          Some(asInt)
+        else
+          None
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toLong: Option[Long] = {
+    if ((constructedFlag & NumberFlags.long) == NumberFlags.long)
+      Some(value.toLong)
+    else {
+      try {
+        val asLong = value.toLong
+        if (BigInt(value) == BigInt(asLong))
+          Some(asLong)
+        else
+          None
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toBigInt: Option[BigInt] = {
+    if ((constructedFlag & NumberFlags.bigInt) == NumberFlags.bigInt)
+      Some(BigInt(value))
+    else {
+      try {
+        Some(BigInt(value))
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toBigDecimal: Option[BigDecimal] = {
+    try {
+      Some(BigDecimal(value))
+    } catch {
+      case _: NumberFormatException => None
+    }
+  }
+
+  def toFloat: Option[Float] = {
+    if ((constructedFlag & NumberFlags.float) == NumberFlags.float)
+      Some(value.toFloat)
+    else {
+      val asFloat = value.toFloat
+      if (BigDecimal(value) == BigDecimal(asFloat.toDouble))
```

Comment:
Getting rid of `Float` for correctness is no more sensible than getting rid of `Double` for correctness.  Also, I am not convinced that possible loss of precision of trailing digits is a valid reason for returning `None` for a floating point value.  For example, `1.0/7` is printed as `0.14285714285714285` but the conventional rounding, which also parses to the same number, is `0.14285714285714286`.

```
0.142857142857142857142857142857142857142857142857142857...
0.142857142857142849212692681248881854116916656494140625
                  ^ stops matching decimal expansion
```
So neither having the exact decimal representation of the double value nor having the best rounding of that value really represents what's going on.  Instead, if you overflow (get positive or negative infinity) that's wrong; and if you underflow (the number is not zero but you return zero) that is also wrong.  Otherwise, the parse should just return the best representation of the value, and everyone should keep in mind (just like always) `Float` and `Double ` might not represent things exactly.

The reason to get rid of `toFloat` is if you want to care about exactness _less_ because you already can't win that battle.